diode/webclient/renderer.js

```diff
@@ -205,6 +205,161 @@ class CanvasManager {
+     * @param {*} sdfg_list         List of SDFGs and nested SDFGs
+     * @param {*} state_parent_list List of parent elements to SDFG states
+     */
+    translate_element(el, x, y, entire_graph, sdfg_list, state_parent_list) {
+        this.stopAnimation();
+
+        // Calculate movement distance with current zoom level
+        let dx = x / this.user_transform.a;
+        let dy = y / this.user_transform.d;
+
+        // Edges connected to the moving element
+        let out_edges = [];
+        let in_edges = [];
+
+        // Find the parent graph in the list of available SDFGs
+        let parent_graph = sdfg_list[el.sdfg.sdfg_list_id];
+        let parent_element = null;
+
+        if (entire_graph !== parent_graph && el.data.state) {
+            // If the parent graph and the entire SDFG shown are not the same,
+            // we're currently in a nested SDFG. If we're also moving a state,
+            // this means that its parent element is found in the list of
+            // parents to states (state_parent_list)
+            parent_element = state_parent_list[el.sdfg.sdfg_list_id];
+        } else if (el.parent_id !== null && parent_graph) {
+            // If the parent_id isn't null and there is a parent graph, we can
+            // look up the parent node via the element's parent_id
+            parent_element = parent_graph.node(el.parent_id);
+            // If our parent element is a state, we want the state's graph
+            if (parent_element.data.state)
+                parent_graph = parent_element.data.graph;
+        }
+
+        if (parent_graph) {
+            // Find all the edges connected to the moving node
+            parent_graph.outEdges(el.id).forEach(edge_id => {
+                out_edges.push(parent_graph.edge(edge_id));
+            });
+            parent_graph.inEdges(el.id).forEach(edge_id => {
+                in_edges.push(parent_graph.edge(edge_id));
+            });
+        }
+
+        if (parent_element) {
+            // Calculate the box to bind the element to. This is given by
+            // the parent element, i.e. the element where out to-be-moved
+            // element is contained within
+            const min_x = (parent_element.x - (parent_element.width / 2)) +
+                (el.width / 2);
+            const min_y = (parent_element.y - (parent_element.height / 2)) +
+                (el.height / 2);
+            const max_x = (min_x + parent_element.width) - el.width;
+            const max_y = (min_y + parent_element.height) - el.height;
+
+            // Make sure we don't move our element outside its parent's
+            // bounding box
+            const target_x = el.x + dx;
+            const target_y = el.y + dy;
+            if (target_x <= min_x || target_x >= max_x) dx = 0;
+            if (target_y <= min_y || target_y >= max_y) dy = 0;
+        }
+
+        if (el.data.type && (el.data.type === 'Memlet' || el.data.type === 'InterstateEdge')) {
```

Comment:
you can check if it's a `typeof Edge` as far as I remember

Reply:
assuming you meant `instanceof`, yes that works, thank you!